core/src/storage/lazy_chunk.rs

```diff
@@ -0,0 +1,393 @@
+                Push::push(value, key_ptr);
+            }
+            None => {
+                // Clear the storage at the given index.
+                crate::env::clear_contract_storage(key_ptr.next_for::<T>());
+            }
+        }
+    }
+}
+
+impl<T> Entry<T> {
+    /// Returns `true` if the cached value of the entry has potentially been mutated.
+    pub fn mutated(&self) -> bool {
+        self.mutated.get()
+    }
+
+    /// Returns a shared reference to the value of the entry.
+    pub fn value(&self) -> Option<&T> {
+        self.value.as_ref()
+    }
+
+    /// Returns an exclusive reference to the value of the entry.
+    ///
+    /// # Note
+    ///
+    /// This changes the `mutate` state of the entry if the entry was occupied
+    /// since the caller could potentially change the returned value.
+    pub fn value_mut(&mut self) -> Option<&mut T> {
+        self.mutated.set(self.value.is_some());
+        self.value.as_mut()
+    }
+
+    /// Takes the value from the entry and returns it.
+    ///
+    /// # Note
+    ///
+    /// This changes the `mutate` state of the entry if the entry was occupied.
+    pub fn take_value(&mut self) -> Option<T> {
+        self.mutated.set(self.value.is_some());
+        self.value.take()
+    }
+
+    /// Puts the new value into the entry and returns the old value.
+    ///
+    /// # Note
+    ///
+    /// This changes the `mutate` state of the entry to `true` as long as at
+    /// least one of `old_value` and `new_value` is `Some`.
+    pub fn put(&mut self, new_value: Option<T>) -> Option<T> {
+        match new_value {
+            Some(new_value) => {
+                self.mutated.set(true);
+                self.value.replace(new_value)
+            }
+            None => self.take_value(),
+        }
+    }
+}
+
+impl<T> LazyChunk<T> {
+    /// Creates a new empty lazy chunk that cannot be mutated.
+    pub fn new() -> Self {
+        Self {
+            key: None,
+            cached_entries: UnsafeCell::new(EntryMap::new()),
+        }
+    }
+
+    /// Performs the given closure on the mutable cached entries.
+    ///
+    /// # Note
+    ///
+    /// Actions on the mutable lazy entries are performed within the closure
+    /// to not leak exclusive references to it to the outside. This is important
+    /// since the `for_entries` method itself operates only on `&self`.
+    fn for_cached_entries<F, R>(&self, f: F) -> R
+    where
+        F: FnOnce(&mut EntryMap<T>) -> R,
+    {
+        // SAFETY: We operate on an exclusive reference on `BTreeMap` within the
+        //         given closure while our method receiver is only a shared reference.
+        //         However, due to encapsulation of the exclusive reference within
+        //         the given closure we cannot leak the exclusive reference outside
+        //         of the closure. So the below action is safe in this regard.
+        f(unsafe { &mut *self.cached_entries.get() })
```

Comment:
Could this code be simultaneously called from several places? Like threads or async functions. How do we protect against such scenario?

Reply:
All ink! code is always executed in a single thread. So no.

Reply:
But, this function should itself be `unsafe` because it is not safe to do 

```
let x = chunk.get(index);
chunk.for_cached_entries(|map| map.remove(&index));
// *x points to dropped data now.
```

Reply:
I think I will simply remove the function since it provides no safe interface.

Reply:
> But, this function should itself be `unsafe` because it is not safe to do
> 
> ```
> let x = chunk.get(index);
> chunk.for_cached_entries(|map| map.remove(&index));
> // *x points to dropped data now.
> ```

Wow ... how could I not see this immediately. O.O